ssh/agent.go

```diff
@@ -0,0 +1,166 @@
+)
+
+// ssh-agent constant identifiers
+const (
+	AgentPidIdentifier = "SSH_AGENT_PID"
+	AuthSockIdentifier = "SSH_AUTH_SOCK"
+)
+
+// singleton instance of agent
+var current *agent
+
+// agentInfo captures the connection information of the ssh-agent
+type agentInfo map[string]string
+
+// agent represents an instance of the ssh-agent
+type agent struct {
+	// pid of the ssh-agent
+	pid string
+
+	// Authentication socket to communicate with the ssh-agent
+	authSockPath string
+
+	// File paths of the keys added to the ssh-agent
+	keyPaths []string
+}
+
+// agentFromInfo parses the information map and returns an instance of agent
+func agentFromInfo(agentInfo agentInfo) *agent {
+	return &agent{
+		pid:          agentInfo[AgentPidIdentifier],
+		authSockPath: agentInfo[AuthSockIdentifier],
+	}
+}
+
+// AddKey adds a key to the ssh-agent process
+func (agent *agent) AddKey(keyPath string) error {
+	e := echo.New()
+	sshAddCmd := e.Prog.Avail("ssh-add")
+	if len(sshAddCmd) == 0 {
+		return errors.New("ssh-add not found")
+	}
+
+	p := e.Env(agent.GetEnvVariables()).
+		RunProc(fmt.Sprintf("ssh-add %s", keyPath))
+	if err := p.Err(); err != nil {
+		return errors.Wrapf(err, "could not add key %s to ssh-agent", keyPath)
+	}
+	agent.keyPaths = append(agent.keyPaths, keyPath)
+	return nil
+}
+
+// Stop removes the keys and kills the ssh-agent process
+func (agent *agent) Stop() error {
+	// set the global variable to nil
+	defer func() {
+		current = nil
+	}()
+
+	e := echo.New()
+	sshAddCmd := e.Prog.Avail("ssh-add")
+	if len(sshAddCmd) == 0 {
+		return fmt.Errorf("ssh-add not found")
+	}
+
+	for _, path := range agent.keyPaths {
+		logrus.Debugf("removing key from ssh-agent: %s", path)
+		p := e.Env(agent.GetEnvVariables()).
+			RunProc(fmt.Sprintf("%s -d %s", sshAddCmd, path))
+		if p.Err() != nil {
+			return fmt.Errorf("failed to remove SSH key from agent: %s: %s", p.Err(), p.Result())
+		}
+	}
+
+	logrus.Debugf("stopping the ssh-agent with pid: %s", agent.pid)
+	e.Env(agent.GetEnvVariables()).Run("ssh-agent -k")
+
+	return nil
+}
+
+// GetEnvVariables returns the space separated key=value information used to communicate with the ssh-agent
+func (agent *agent) GetEnvVariables() string {
+	return fmt.Sprintf("%s=%s %s=%s", AgentPidIdentifier, agent.pid, AuthSockIdentifier, agent.authSockPath)
+}
+
+// GetAgent returns the current singleton instance of the ssh-agent
+func GetAgent() *agent {
+	return current
+}
+
+// StartAgent starts the ssh-agent process and returns the SSH authentication parameters.
+func StartAgent() (*agent, error) {
+	p := echo.New().RunProc("ssh-agent -s")
+	if p.Err() != nil {
+		return nil, errors.Wrap(p.Err(), "failed to start ssh agent")
+	}
+
+	agentInfo, err := parseAgentInfo(p.Out())
+	if err != nil {
+		return nil, err
+	}
+	if err := validateAgentInfo(agentInfo); err != nil {
+		return nil, err
+	}
+
+	// populate the singleton reference of the agent
+	current = agentFromInfo(agentInfo)
```

Comment:
Just for precaution sake, let's add a write Mutex here when updating current.

Reply:
Done